File: leetcode_javascript/762. Prime Number of Set Bits in Binary Representation.js
/**
 * @param {number} left
 * @param {number} right
 * @return {number}
 */
// 1.
// var countPrimeSetBits = function (left, right) {
//   let res = 0;
//   let visited = {};
//   for (let i = left; i <= right; i++) {
//     let ones = i
//       .toString(2)
//       .split("")
//       .filter((el) => el === "1").length;
//     if (visited[ones] === undefined) visited[ones] = isPrime(ones);
//     if (visited[ones]) res++;
//   }
//   return res;

//   function isPrime(n) {
//     if (n === 1) return false;
//     for (let i = 2; i < n; i++) {
//       if (n % i === 0) return false;
//     }
//     return true;
//   }
// };

// 2.
var countPrimeSetBits = function (left, right) {
  let primes = new Set([2, 3, 5, 7, 11, 13, 17, 19]);
  let res = 0;

  for (let i = left; i <= right; i++) {
    if (primes.has(getOnes(i))) res++;
  }
  return res;

  function getOnes(n) {
    let res = 0;
    while (n) {
      if (n & 1) res++;
      n >>>= 1;
    }
    return res;
  }
};
// [2, 3, 5, 7, 11, 13, 17, 19];
let left = 6,
  right = 10;
Output: 4;
console.log((10 ** 6).toString(2).length);
console.log(countPrimeSetBits(left, right));

File: c/762_Prime_Number_of_Set_Bits_in_Binary_Representation.cpp
#include <iostream>
using namespace std;

class Solution {
 public:
  int countPrimeSetBits(int left, int right) {
    int primes = 0;
    for (int i = left; i <= right; i++) {
      int ones = 0;
      int n = i;
      while (n) {
        if (n & 1) ones++;
        n >>= 1;
      }
      if (isPrime(ones)) primes++;
    }
    return primes;
  }

  bool isPrime(int n) {
    if (n == 1) return false;
    for (int i = 2; i < n; i++) {
      if (n % i == 0) return false;
    }

    return true;
  }
};

int main(int argc, char const *argv[]) {
  Solution s;
  int left = 6;
  int right = 10;

  cout << s.countPrimeSetBits(left, right) << endl;
  return 0;
}
